### ISSUE TITLE
[Question] Common settings.yml for organisation and conflicts with the auto-detected settings for a consumer?

### ISSUE THREAD

**Issue by Philippe-Cote-Morneault** [conan-io/conan#5167]
To help us debug your issue please explain:

Hi, I am having a hard time understanding the behavior of doing `conan config install` to specify `settings.yml` and then auto-detection of settings by Conan. Here's what I want to do: I want all consumers of my organisation to not have to interact with Conan directly. I am making a script that installs Conan and deals with the configuration for them. These consumers may have Windows, MacOS, or Linux. However, I want to only allow them to have one compiler version available for each platform. For instance, Windows consumers must only use Visual Studio 2015. I want to specify in `settings.yml` what version of compiler is possible for each platform, but I fear that it could cause some sort of conflict if the auto-detected settings are not the one specified in the `settings.yml`. 

The documentation states that the settings for the default profile are auto-detected the first time a Conan command is executed. In my case, the first command that will be run is `conan config install`. If I understand, it is possible that the settings are auto-detected before Conan gets the configuration. Let's say that my consumer has Visual Studio 2015, 2017 and 2019. It might detect that his compiler version is the one of VS 2017. Then, when the settings get set and only allow VS version 2015. Will this cause an issue since that it will detect my consumer's compiler version is not specified in `settings.yml`? I basically have a hard time understanding how Conan deals with a consumer having multiple compiler version when auto-detecting and the conflicts that may arise when using a `settings.yml` for my organisation that doesn't include the compiler that got auto-detected at first. 

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [X] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by Johnnyxy** [comment]
Hi there,
this will not answer if your case would create an issue with Conan. Instead I want to point out some issues you will have with this approach in general.

What you want to do is not the use case of Conan. Conan is a package management system. What you are trying to do is to utilize it as some sort of software deployment / management system.

If you want to restrict the users then do not allow them to install anything. Like for Visual Studio one needs administrative rights to do that. But sadly under windows developing without administrative rights can be quite a challenge sometimes (depending on the type of software one develops).

Even restricting the compiler/software to use is not 100% possible. As there are enough toolchains out there which runs out-of-the-box without any installation (just copy; portable). What you would have to restrict the use of programms with some monitoring system.

As the ```settings.yml``` resides in the a user-controlled environment you cannot force the user to use some specific settings at all time. Your init-scripts which run and update the Conan config do not run all the time. Between those executions one can always modify the config and do what he/she wants with Conan. Or one can just use some other location to manage his Conan home directory ([see here](https://docs.conan.io/en/latest/reference/env_vars.html#conan-user-home)).

If you want to control the consumer systems then you should have a look for some configuration management system (e.g. [Ansible](https://en.wikipedia.org/wiki/Ansible_(software)), [Puppet](https://en.wikipedia.org/wiki/Puppet_(software)), etc.).
With Ansible for example you could uninstall a Visual Studio automatically if it is found.

**Comment by uilianries** [comment]
> However, I want to only allow them to have one compiler version available for each platform.

Why you just don't provide some profiles instead? Actually you can't prevent people changing the settings, anyone is able edit settings.yml and add a new compiler version.

> The documentation states that the settings for the default profile are auto-detected the first time a Conan command is executed. In my case, the first command that will be run is conan config install.

Not exactly, Conan will generate the default profile only when it needs a profile e.g. `conan create` and there is no profile. Running `conan config install` will populate the profiles folder, but it won't generate a default profile if your remote config doesn't provide one.

> If I understand, it is possible that the settings are auto-detected before Conan gets the configuration.

You can force the default profile generation: `conan profile new --detect default`

> Let's say that my consumer has Visual Studio 2015, 2017 and 2019. It might detect that his compiler version is the one of VS 2017. Then, when the settings get set and only allow VS version 2015. Will this cause an issue since that it will detect my consumer's compiler version is not specified in settings.yml?

To detect the profile Conan will work fine, but when building some project, an error will came up about invalid "settings.compiler.version".

> As the settings.yml resides in the a user-controlled environment you cannot force the user to use some specific settings at all time

But if you provide a default settings.yml used by all organization, and if an user doesn't want to use your config, their is using by own risk. I worked in company with many different configurations, from server to embedded stuff. People there were able to customize organization config by PR, including new profiles, compiler versions, or they could fork and restrict to the team.

 > If you want to control the consumer systems then you should have a look for some configuration management system (e.g. Ansible, Puppet, etc.).

Yes, Ansible, Puppet, Chef are amazing, but they are another type of tool for another problem.

To remove compiler versions, you will need to detect which compiler versions are installed and update settings.yml. You could write a Conan hook to do this.

IMO you won't be able to restrict your users, anyone is able to edit the settings file. Most of time people use custom settings to add new cpus and compiler versions. I would say to provide profiles instead of burning your brain solving it, probably users will complain you when some compiler is not listed, on the other hand they could use a specific profile. You can also provide a hook to check which compiler should be used and update the default profile.

Regards.

**Comment by JigsawCorp** [comment]
@Johnnyxy Thank you for your detailed response!

Reading back my first comment, I realize that I didn't properly describe our use case. We in fact do not really use Conan to it's potential, we mostly just want to be able to remotely host our binary libraries to a certain remote. We do however want to support multiple settings. I gave a very narrow example, but what I mean is that it is possible that we would have some packages on our Conan for VS 2013, 2015 and 2017, but not 2019. Let's say I have a dev on the team that wants to pull our git repo and start building the project, but he has VS 2013, 2015, 2017 and 2019 on his computer. I want my `settings.yml `to have a list of all supported packages versions for VS, so I would have 2013, 2015 and 2017. I'm wondering if there is a danger that the auto-detected settings of the dev would find VS 2019 since he has it installed, try to get a package that has a version for VS 2019, not find it and then return an error? 

Basically, if I have a dev that has multiple compilers on his computer, having a few that support a package on the remote and a few others that do not, I want him to download a package for a version of a compiler that he has and then build my project with that. My script can find out what version of VS the package was downloaded for and then build a VS project of that version, but I'm afraid that the Conan part will have a hard time dealing with auto-detecting a compiler that has a matching package on our remote.

Would this be a proper use case of Conan? If so, what exactly would happen in the scenario I provided?

**Comment by JigsawCorp** [comment]
@uilianries Thanks for the reply!

As I stated in the comment right above this one, I gave a very narrow example as a sort of edge case. However, our most likely scenario is that we would have versions for let's say VS 2013, 2015 and 2017, but not 2019. In fact, we could have the same binaries be compatible with 2013, 2015 and 2017, but just not with 2019 and we would be to busy to spend resources to change the binaries to have a 2019 version. However, one of our devs could have VS 2013, 2015, 2017 and 2019 on his computer. We want to make sure that when he tries to build the project, it gets built in a version that has packages on the remote. From my understanding, the point of setting.yml was to specify what settings we allow our consumers to have in order to have a package available for this consumer's configuration. As I said, I can handle in my script building the project for the right version, but I'm afraid that Conan will auto-detect VS 2019, try to fetch a package for VS 2019 and fail since it cannot find one. I want my consumer to be able to fetch an existing package for settings that he has, but make sure that Conan chooses the settings that he both has and a package exists for.

**Comment by Johnnyxy** [comment]
Ouh, I think you are resembling the use case of Conan here :).

You want the dev to be able to pull a package that is available on the remote for a compiler version he has? This is already a functionality conan possesses. As consumer one has to provide the compiler that he want the packages to pull for and conan pulls just the right one. If no package is available for that compiler then conan tries to build it or just stops with an error.

If you do not want Conan to be able to build your software recipe or you want to restrict it to the compilers you support you can do this in the recipe's ```def configure(self)``` function ([see here](https://docs.conan.io/en/latest/reference/conanfile/methods.html#configure-config-options) and [here](https://docs.conan.io/en/latest/mastering/conditional.html?highlight=conaninvalidconfiguration#constrain-settings-and-options)).

This way noone can use your recipe and its packages without the correct compilers.

**Comment by JigsawCorp** [comment]
@Johnnyxy 

> Ouh, I think you are resembling the use case of Conan here :).

You had me worried that I spent a lot of time learning Conan for no good reason haha

> As consumer one has to provide the compiler that he want the packages to pull for and conan pulls just the right one.

Well my main question is how Conan auto-detects the settings for the user. If I have 5 versions of Visual Studio on my computer, how does Conan select one for the default profile? Technically, we do have to provide the version of the compiler to Conan, but implicitly, if we do not use the -`s ` options, Conan will try with the default profile, that would have been auto-detected. If these auto-detected settings are not some that have a corresponding package, Conan will just fail and that's it. I would like to know if there is a way to write all possible settings we support in `settings.yml`, have the user do a `conan config install`, and then Conan set a default profile to match settings that both the user has, and settings contains. 

Basically, I don't want my user to go set his profile himself. I want to have some sort of way, whether it is handled completely my Conan, or heavily assisted by scripts, to look through possible settings in `settings.yml` and find a configuration the user has on his machine and use these settings if the user does have these settings.

**Comment by Johnnyxy** [comment]
Ah I begin to understand what you want. Yes if Conan detects for the consumer a compiler version that you do not support but if the consumer has a compatible compiler installed Conan would still complain that the current setting cannot be applied to your recipe.
What you want is to provide an alternative to the consumer that he can get a package for some other compiler/version.

But it does not work this way. One has to provide Conan at invocation with the compiler version one wants. Then Conan can do its magic. But if one provides an invalid configuration for your recipe then it would be without sense anyway in this particular Conan call to pull something different. This would be an inconsistent state.

To do what you want you have to do some logic before invocing Conan. You can query the remote what variants are available and then do some computation what to pull and then invoke Conan with the right settings.
But this would be outside of Conan's domain :).

**Comment by JigsawCorp** [comment]
@Johnnyxy 

Perfect, that does answer my question. 

I don't know how feasible it would be, but it would be a neat feature to have Conan do what I just described. The auto-detection has the possibility of detecting any compiler installed. Chances that this compiler doesn't have an associated package on a remote are quite high. This means that we are always going to need to manually set our settings our profile as a consumer. This renders the auto-detection feature somewhat useless from my perspective. If Conan could try to find a combination of settings of the user and of the package and install that package, it would be great. I'm not too sure how this would be implemented nor the general logic of such a thing. I doubt that I am the only one who would benefit from Conan handling this. Somehow matching the CMake generator to have variables with the settings that the packages have been downloaded with would allow us to build the proper solution in CMake without some complicated logic, just variables that your CMake generator would do. 

Anyways, consider this as a suggestion. I know it's a big feature, but I bet this would be very useful!

**Comment by lasote** [comment]
I think you might want to distribute the valid profiles also with `conan config install` and then, depending on the platform, create a `default` profile just including the right one. e.g:

```
include(vs15)
```

**Comment by memsharded** [comment]
Closing this issue as outdated. Many things changed since then, for some of the above side issues, now the ``compatibility.py`` plugin exist in Conan 2.0, that allows to define custom compatibility of binaries if desired (like compatibility among different VS versions). This might still not be recommended in many cases, and alignment of using the same ``profiles`` distributed with ``conan config install`` is probably the recommended approach.
